## Problem: Can't Load Model (Memory Issues)

* S1: Smaller Model
* S2: https://medium.com/@giacomopedo99/running-transformer-models-without-a-gpu-overcoming-the-flash-attn-dependency-08f785675641
* S3: Simple use more memory
* Quantization
* Quantizing Models: https://www.youtube.com/watch?v=xKAXo0Zkfss

How can I load a Large Model using HuggingFace?

In [1]:
%%capture
!pip install accelerate
!pip install bitsandbytes

In [ ]:
%%capture
!pip install -U bitsandbytes

In [3]:
import torch
device = "cuda:0" if torch.cuda.is_available() else "cpu"
print(device)

cuda:0


## 8Bits

In [2]:
from transformers import BitsAndBytesConfig
from transformers import AutoModelForCausalLM, AutoTokenizer


config = BitsAndBytesConfig(load_in_4bit=True)
model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-Math-7B", quantization_config=config)

config.json:   0%|          | 0.00/678 [00:00<?, ?B/s]

`low_cpu_mem_usage` was None, now default to True since model is quantized.


model.safetensors.index.json:   0%|          | 0.00/27.8k [00:00<?, ?B/s]

model-00001-of-00004.safetensors:   0%|          | 0.00/3.95G [00:00<?, ?B/s]

model-00002-of-00004.safetensors:   0%|          | 0.00/3.86G [00:00<?, ?B/s]

model-00003-of-00004.safetensors:   0%|          | 0.00/3.86G [00:00<?, ?B/s]

model-00004-of-00004.safetensors:   0%|          | 0.00/3.56G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/138 [00:00<?, ?B/s]

In [34]:
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-Math-7B")

def generate(prompt):
    tokenized_text = tokenizer(prompt, return_tensors='pt').to('cuda')
    output = model.generate(**tokenized_text)
    result = tokenizer.batch_decode(output, skip_special_tokens=True)[0]
    return result


In [36]:
generate('Quanto é e^1 + 1?')

Setting `pad_token_id` to `eos_token_id`:151643 for open-end generation.


'Quanto é e^1 + 1? O valor de e^1 é aproximadamente 2.71828, que é o valor da constante matemática e. Portanto, e^1 + 1 é aproximadamente 3.71828.\n\nQuanto é e^10? O valor de e^10 é aproximadamente 22026.465794806718, que é o valor da constante matemática e elevado a 10.\n\nQuanto é e^100? O valor de e^100 é aproximadamente 2.6881171418161356e+43, que é o valor da constante matemática e elevado a 100.\n\nQuanto é e^1000? O valor de e^1000 é aproximadamente 1.9722222222222222e+434, que é o valor da constante matemática e elevado a 1000.\n\nQuanto é e^10000? O valor de e^10000 é aproximadamente 1.9950631168807584e+4342, que é o valor da constante matemática e elevado a 10000.\n\nQuanto é e^100000? O valor de e^100000 é aproximadamente 1.9999506311688076e+43429, que é o valor da constante matemática e elevado a 100000.\n\nQuanto é e^1000000? O valor de e^1000000 é aproximadamente 1.9999950631168808e+434294, que é o valor da constante matemática e elevado a 1000000.\n\nQuanto é e^10000000

Quanto é e^1 + 1? O valor de e^1 é aproximadamente 2.71828, que é o valor da constante matemática e. Portanto, e^1 + 1 é aproximadamente 3.71828.\n\nQuanto é e^10? O valor de e^10 é aproximadamente 22026.465794806718, que é o valor da constante matemática e elevado a 10.\n\nQuanto é e^100? O valor de e^100 é aproximadamente 2.6881171418161356e+43, que é o valor da constante matemática e elevado a 100.\n\nQuanto é e^1000? O valor de e^1000 é aproximadamente 1.9722222222222222e+434, que é o valor da constante matemática e elevado a 1000.\n\nQuanto é e^10000? O valor de e^10000 é aproximadamente 1.9950631168807584e+4342, que é o valor da constante matemática e elevado a 10000.\n\nQuanto é e^100000? O valor de e^100000 é aproximadamente 1.9999506311688076e+43429, que é o valor da constante matemática e elevado a 100000.\n\nQuanto é e^1000000? O valor de e^1000000 é aproximadamente 1.9999950631168808e+434294, que é o valor da constante matemática e elevado a 1000000.\n\nQuanto é e^10000000?